### PR TITLE
adding concurrent request capabilities as well as named routing

### PR DESCRIPTION
**Pull request by alanvoss**
We decided to privately fork until my changes are finished.  Here's the other PR: https://github.com/PSPDFKit-labs/bypass/pull/36

Basically, this adds routing, concurrency, and counted route hits.  See the other PR for more info.

This is what the README will look like once merged:
https://github.com/alanvoss/bypass/blob/support_multi_expect_attempt2/README.md

### CODE REVIEW

File: lib/bypass/instance.ex [hunk 1]
```diff
@@ -79,45 +91,190 @@ defmodule Bypass.Instance do
...
+  defp never_problems(expectations) do
+    expectations
+    |> filter_problems_by_expected(:never, fn {_, count} -> count > 0 end)
+    |> Enum.map(fn
+         {{:any, :any}, count} ->
+           count_problem("bypass", "to never be called", count)
+         {{method, path}, count} ->
+           count_problem("#{method} #{path}", "to never be called", count)
+       end)
+  end
```
Comment (Christopher-Bui) [inline lib/bypass/instance.ex]: Possible to DRY these up 3 `..._problems` functions up?

File: lib/bypass/instance.ex [hunk 2]
```diff
@@ -79,45 +91,190 @@ defmodule Bypass.Instance do
...
+        retained_plugs: Map.delete(plugs, ref),
+        results: [result | Map.fetch!(route_expectations, :results)]
+      })
+    end)
+  end
+
+  defp expect_problem_messages(expectations) do
+    once_problems(expectations)
+    |> List.flatten(once_or_more_problems(expectations))
+    |> List.flatten(never_problems(expectations))
```
Comment (Christopher-Bui) [inline lib/bypass/instance.ex]: Why not just `List.flatten([once_problems(expectations), one_or_more_problems(expectations)...]`

File: lib/bypass/instance.ex [hunk 3]
```diff
@@ -79,45 +91,190 @@ defmodule Bypass.Instance do
...
         %{socket: nil} -> state
         %{socket: socket, ref: ref} ->
           do_down(ref, socket)
           %{state | socket: nil}
       end
-    {:stop, :normal, state.request_result, updated_state}
+
+    result = 
+      cond do
+        state.pass == true ->
```
Comment (Christopher-Bui) [inline lib/bypass/instance.ex]: Is truthy `state.pass` good enough or do you need it to be `true`?

File: lib/bypass/instance.ex [hunk 4]
```diff
@@ -56,12 +56,24 @@ defmodule Bypass.Instance do
...
+      )
     ]
-    {:noreply, Map.update!(state, :retained_plug, fn nil -> caller_pid end)}
+
+    updated_state =
+      update_in(state[:expectations][route][:retained_plugs], fn plugs ->
+        Map.update(plugs, ref, caller_pid, fn _ ->
+          raise ExUnit.AssertionError, "plug already installed for #{method} #{path}"
+        end)
+      end)
```
Comment (alanvoss) [inline lib/bypass/instance.ex]: this version:
https://hexdocs.pm/elixir/master/Kernel.html#update_in/2

it's the macro.  i think it's easier to read.

File: test/bypass_test.exs [hunk 5]
```diff
@@ -140,17 +234,238 @@ defmodule BypassTest do
     assert_received ^ref
   end
 
+  @tag :wip
```
Comment (alanvoss) [inline test/bypass_test.exs]: good catch.

File: lib/bypass/instance.ex (hunk 3 above)
Comment (alanvoss) [inline lib/bypass/instance.ex]: good call.

File: lib/bypass/instance.ex (hunk 2 above)
Comment (alanvoss) [inline lib/bypass/instance.ex]: yep.  again.

File: lib/bypass/instance.ex (hunk 1 above)
Comment (alanvoss) [inline lib/bypass/instance.ex]: agreed.  this was gross and I knew it.  thanks for agreeing.  DRYed out.

**Comment by alanvoss** [comment]
merging.

**Pull request merged**